src/batch/src/executor/join/hash_join.rs

```diff
@@ -360,15 +365,25 @@ impl BoxedExecutorBuilder for HashJoinExecutorBuilder {
         ensure!(params.left_key_columns.len() == params.right_key_columns.len());
 
         let hash_key_kind = calc_hash_key_kind(&params.right_key_types);
-
+        let output_indices: Vec<usize> = hash_join_node
```

Comment:
What if the `output_indices` is empty?

Reply:
Sorry for the delay and I just dig into the code a little bit. :rofl: 

IIUC, when `output_indices` is empty, it much resembles an empty project node. In that case, only the cardinality is kept, and the `Vec<Column>` is empty. And for stream executors, the `ops` is kept as well.

The implementation in this PR heavily relies on the `reorder_columns` method of `DataChunk` and `StreamChunk`. I've checked them([[1]](https://github.com/singularity-data/risingwave/blob/3289e2b8c6722330d92db2acbf628bf8a574dc9e/src/common/src/array/stream_chunk.rs#L275-L280), [[2]](https://github.com/singularity-data/risingwave/blob/3289e2b8c6722330d92db2acbf628bf8a574dc9e/src/common/src/array/data_chunk.rs#L425)) and, to my surprise, I found that they do keep `ops` and `visibility` even if all columns are removed. So I think this pr may correctly handle this situation. I will add some tests maybe. And please tell me if I'm missing something.

Reply:
Another problem is what if we don't need to prune any column? e.g. Keep all columns.

Reply:
In current code when no columns are required to be reordered, `reorder_column` will still be called and do 'reordering'. And this may have some overhead. Maybe I can check whether the reordering is needed before actually performing reorder. So we can avoid the overhead.

Reply:
Yes, since this is a per chunk operation, it may cause much overhead. I would suggest to check this.

Reply:
actually, `reorder_column` is completely unecessary, we should have the chunkbuilder do the reordering by being aware of output indices. Because the chunkbuilder is cloning the Datum for each value in the row already.

@Enter-tainer , maybe I should have been more clear about what I meant! Do you think it could be improved in a follow up PR?

Reply:
> Do you think it could be improved in a follow up PR?

Of course!

I just don't want this PR being to big and hard to review🤣, so I dicide to do it in following PRs.